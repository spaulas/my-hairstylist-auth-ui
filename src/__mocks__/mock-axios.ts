export const mockPost = jest.fn().mockReturnValue({
  data: {},
  callback: jest.fn(),
});

export const mockGet = jest.fn().mockReturnValue({
  data: {},
  callback: jest.fn(),
});

export const mockPut = jest.fn().mockReturnValue({
  data: {},
  callback: jest.fn(),
});

export const mockDelete = jest.fn().mockReturnValue({
  data: {},
  callback: jest.fn(),
});

export const mockPatch = jest.fn().mockReturnValue({
  data: {},
  callback: jest.fn(),
});

const mockAxios = jest.fn().mockReturnValue({
  post: mockPost,
  get: mockGet,
  put: mockPut,
  patch: mockPatch,
  delete: mockDelete,
});

jest.mock("../config/HttpClient/index.ts", () => ({
  __esModule: true,
  default: mockAxios,
}));
