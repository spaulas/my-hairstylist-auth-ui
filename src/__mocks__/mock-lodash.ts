export const mockDebounce = jest.fn((fn) => console.log('MOCK DEBOUTNCE'));

jest.mock("lodash.debounce", () => ({
  ...jest.requireActual("lodash.debounce"),
  debounce: mockDebounce,
}));
