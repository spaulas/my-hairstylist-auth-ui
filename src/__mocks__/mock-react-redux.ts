import { ExplicitAny } from "@type/global";

export const mockSelector = jest.fn();

export const mockDispatch = jest.fn().mockImplementation((action: ExplicitAny) => null);

jest.mock('react-redux', () => ({
    ...jest.requireActual('react-redux'),
    useSelector: mockSelector,
    useDispatch: () => mockDispatch
}));
