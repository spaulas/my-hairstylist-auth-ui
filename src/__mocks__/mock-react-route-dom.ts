export const mockPush = jest.fn();
export const mockBlock = jest.fn();
export const mockLocation = jest.fn();

export const mockParams = jest.fn().mockImplementation(() => {});

jest.mock('react-router-dom', () => ({
    useLocation: mockLocation,
    useParams: mockParams,
    useHistory: () => ({
        push: mockPush,
        block: () => ({
            current: mockBlock
        })
    })
}));
