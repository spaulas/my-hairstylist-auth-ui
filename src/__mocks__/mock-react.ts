export const mockUseState = jest.fn();
export const mockUseCallBack = jest.fn((fn) => fn);

jest.mock("react", () => ({
  ...jest.requireActual("react"),
  useState: mockUseState,
  useCallback: mockUseCallBack,
}));
