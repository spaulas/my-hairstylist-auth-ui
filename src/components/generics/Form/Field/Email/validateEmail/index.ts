const validateEmail = (email: string): boolean => {
  const isValid = /^\w+([\.-]?\w+)*@\w+([\.-]?\w+)*(\.\w{2,3})+$/.test(email);

  return isValid;
};

export default validateEmail;
