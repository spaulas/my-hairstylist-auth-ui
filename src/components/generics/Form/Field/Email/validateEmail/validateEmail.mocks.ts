export const mockValidateEmail = jest.fn();

jest.mock("./index.ts", () => mockValidateEmail);
