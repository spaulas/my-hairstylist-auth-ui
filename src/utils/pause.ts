export default function pause(seconds: number) {
  return new Promise((r) => setTimeout(r, seconds * 1000));
}
