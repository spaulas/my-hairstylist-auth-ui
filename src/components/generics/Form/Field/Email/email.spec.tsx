import "@testing-library/jest-dom/extend-expect";
import { mockUseState } from "../../../../../__mocks__/mock-react";
import { mockValidateEmail } from "./validateEmail/validateEmail.mocks";
import EmailField from ".";
import { act, fireEvent, render, screen } from "@testing-library/react";
import pause from "../../../../../utils/pause";

describe("EmailField", () => {
  const props = {
    email: "",
    setEmail: jest.fn(),
    isFocused: false,
    setIsFocused: jest.fn(),
  };

  const hasError = false;
  const setHasErrorMock = jest.fn();

  const changeEmailInput = async (newEmail: string, isValid: boolean) => {
    mockValidateEmail.mockReturnValue(isValid);

    const emailProps = {
      ...props,
      email: newEmail,
    };
    const { rerender } = render(<EmailField {...props} />);

    const input = screen.getByTestId("email");
    fireEvent.change(input, { target: { value: newEmail } });

    expect(props.setEmail).toHaveBeenCalledWith(newEmail);

    await act(async () => {
      await pause(1);
    });

    expect(mockValidateEmail).toHaveBeenCalledWith(newEmail);
    expect(setHasErrorMock).toHaveBeenCalledWith(!isValid);

    rerender(<EmailField {...emailProps} />);

    expect(screen.getByText("Email")).toBeVisible();
    expect(input).toHaveValue(newEmail);
  };

  beforeEach(() => {
    mockUseState.mockReturnValue([hasError, setHasErrorMock]);
  });

  it("renders correctly", () => {
    const { container } = render(<EmailField {...props} />);
    expect(container).toMatchSnapshot();
  });

  it("renders label when there is no email", () => {
    render(<EmailField {...props} />);
    expect(screen.getByText("Email")).toBeVisible();
  });

  it("renders label and email when there is an email", async () => {
    const email = "email@email.com";
    await changeEmailInput(email, true);
  });

  it("shows error message when email is invalid", async () => {
    const email = "email";
    await changeEmailInput(email, false);

    const errorMessage = screen.getByText("Please add a valid email");
    expect(errorMessage).toBeVisible();
  });
});
