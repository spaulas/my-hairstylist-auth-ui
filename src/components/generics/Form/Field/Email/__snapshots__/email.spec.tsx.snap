// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmailField renders correctly 1`] = `
<div>
  <div
    class="form--item"
  >
    <p
      class="form--label "
    >
      Email
    </p>
    <input
      class="form--input "
      data-testid="email"
      id="email"
      name="email"
      type="email"
      value=""
    />
    <p
      class="error "
    >
      Please add a valid email
    </p>
  </div>
</div>
`;
